### PR TITLE
feat: add new rdkit worker actions

### PR DESCRIPTION
**Pull request by atfera**
Added new actions for rdkit worker:
- isValidMolblock: to validate molblock
- convertMolNotation: to convert mol structure to another notation (will add more notation in the future if necessary)
- getQMolSmarts: to get smarts from a query mol. This can be different from normal mol get_smarts.
-  addHs/removeHs: add or remove explicit Hydrogens. Necessary functions for ketcher
- getNewCoords: get new coord for structure using or not coordgen

### CODE REVIEW

File: src/worker/utils/chem.ts [hunk 1]
```diff
@@ -136,6 +136,99 @@ export const getMatchingSubstructure = ({ structure, substructure }: { structure
...
+  useQMol = false,
+}: {
+  moleculeString: string;
+  targetNotation: MolNotation;
+  sourceNotation?: SourceMolNotation;
+  useQMol?: boolean;
+}): string | null => {
+  if (sourceNotation != null) {
+    if (sourceNotation === targetNotation) throw new Error('Source and target notations must differ');
+    if (!_validateSource(moleculeString, sourceNotation)) throw new Error('Molecule string not valid');
```
Comment (RamziWeslati) [inline src/worker/utils/chem.ts:167]: We should create custom error for the package, or at least append [@iktos-oss/rdkit-provider] to the message, so that it easily identified and in the case of the custom error, we can easily catch/handle it

File: src/worker/utils/chem.ts [hunk 2]
```diff
@@ -136,6 +136,99 @@ export const getMatchingSubstructure = ({ structure, substructure }: { structure
...
+  }
+  const mol = useQMol
+    ? globalThis.workerRDKit.get_qmol(moleculeString)
```
Comment (RamziWeslati) [inline src/worker/utils/chem.ts:170]: are qmols used frequently? we might want to cache them like we are caching mols, and do a `create_qmol` `release_qmol`. But we can do this in another pr if you want

File: src/worker/utils/chem.ts [hunk 3]
```diff
@@ -136,6 +136,99 @@ export const getMatchingSubstructure = ({ structure, substructure }: { structure
...
+  } finally {
+    useQMol ? mol?.delete() : release_molecule(mol);
+  }
+};
+
+export const getNewCoords = (structure: string, useCoordGen?: boolean) => {
+  const mol = get_molecule(structure, globalThis.workerRDKit);
+  if (!mol) return null;
+  try {
+    const mdl = useCoordGen != null ? mol.get_new_coords(useCoordGen) : mol.get_new_coords();
```
Comment (RamziWeslati) [inline src/worker/utils/chem.ts:187]: was this meant to be `useCoordGen !== undefined`? or `export const getNewCoords = (structure: string, useCoordGen?: boolean | null = null) => {` ?

File: src/worker/utils/chem.ts (hunk 2 above)
Comment (atfera) [reply src/worker/utils/chem.ts:172]: Only when using smarts. I don't think it's necessary to use cache for qmol right now.

File: src/worker/utils/chem.ts (hunk 3 above)
Comment (atfera) [reply src/worker/utils/chem.ts:189]: `useCoordGen != null` check if `useCoordGen` is not `null` or `undefined`.

File: src/worker/utils/chem.ts [hunk 4]
```diff
@@ -136,6 +136,101 @@ export const getMatchingSubstructure = ({ structure, substructure }: { structure
...
+  try {
+    return mol.is_valid();
+  } finally {
+    release_molecule(mol);
+  }
+};
+
+/**
+ * Convert molecule structure to given notation (smiles, smarts, molblock, ...)
+ * TODO some notations like inchi don't work with qmol
```
Comment (RamziWeslati) [inline src/worker/utils/chem.ts:152]: is this still TODO or is this comment stale?

**Review by RamziWeslati — DISMISSED** [review]

File: src/worker/utils/chem.ts (hunk 3 above)
Comment (atfera) [reply src/worker/utils/chem.ts]: Yeah I did it like that to avoid passing a null and throw an error, but I just check it and `get_new_coords` seems working fine with `null` and `undefined`.
Changed to `useCoordGen !== undefined`

File: src/worker/utils/chem.ts (hunk 4 above)
Comment (atfera) [reply src/worker/utils/chem.ts]: Changed to warning. We just need to avoid using qmol to convert to aromatic_form (throw error) and inchi (return empty string). I will try to solve this using TS in the future.

**Review by RamziWeslati — APPROVED** [review]
